File: Backjoon/01005_ACMCraft/main.py
DEBUG = False 
def log(s):
  if DEBUG == True:
    print(s)

T = int(input())
for t in range(0, T):
  N, K = map(lambda x: int(x), input().split())
  timeToConstruct = list(map(lambda x: int(x), input().split()))
  orderToConstruct = {}
  for K in range(0, K):
    lhs, rhs = map(lambda x: int(x), input().split())
    if rhs - 1 in orderToConstruct: # lhs to rhs 
      orderToConstruct[rhs - 1].append(lhs - 1)
    else:
      orderToConstruct[rhs - 1] = [lhs - 1]
  toConstrcut = int(input())

  cache = {} # the maximum required time when n(depth)
  st = [toConstrcut - 1]
  while len(st) != 0:
    front = st[0]
    log('front: {} cache? {}'.format(front, front in cache))
    isPop = False 
    if not front in orderToConstruct: # start 
      log('there is no where to go: {}'.format(front))
      cache[front] = timeToConstruct[front]      
    else:
      maxValue = 0 
      for lhs in orderToConstruct[front]:
        log('lhs: {}'.format(lhs))
        if lhs in cache:
          maxValue = max(maxValue, cache[lhs] + timeToConstruct[front])
        else:
          isPop = True
          st.insert(0, lhs)
          break 
      if isPop == False: 
        cache[front] = maxValue
    if isPop == False: 
      st.pop(0)
    log(st)
    log(cache)

  print(cache[toConstrcut - 1])

File: Backjoon/01005_ACMCraft/main.cpp
#include <iostream>
#include <vector>
#include <utility>
#include <algorithm>
#include <stack>
#include <queue>
#include <unordered_set>

using namespace std;

typedef unsigned int ui;
typedef pair<ui, ui> pii;

#define DEBUG
#ifdef DEBUG
#define D(x) x
#else
#define D(x)
#endif


int main(void) {


  ui T;
  cin >> T;

  for(ui i = 0; i < T; i++) {
    ui N, K;
    cin >> N >> K;
    vector<ui> constTime;
    for(ui j = 0; j < N; j++) {
      ui tmp;
      cin >> tmp;
      constTime.emplace_back(tmp);
    }

    vector<pii> steps;
    for(ui j = 0; j < K; j++) {
      ui prev, next;
      cin >> prev >> next;

      steps.push_back(pii(prev - 1, next - 1));
    }
    
    sort(steps.begin(), steps.end(), [](const pii& lhs, const pii& rhs) {
      return lhs.second < rhs.second;
    });

    ui W;
    cin >> W;
    W -= 1;

    ui building = W;
    ui totalTime = constTime[W];
    queue<pii> buildingToBuild;
    unordered_set<ui> setToCheck;

    buildingToBuild.push(pii(W, 0));

    auto findToBuild = [&](ui building, ui idx) -> ui {
      auto findRequired = [&building](const pii& pair) {
        return pair.second == building;
      };
      ui max = 0;
      auto required = lower_bound(steps.begin(), steps.end(), building, [](const pii& lhs, const ui rhs) {
        return lhs.second < rhs;
      });
      while(required != steps.end() && required->second == building) {
        D(cout << "Found: " << required->first << " " << required->second << endl;)
        if(setToCheck.find(required->first) == setToCheck.end()) {
          D(cout << "Push to Set: " << required->first << endl;)
          setToCheck.insert(required->first);
          buildingToBuild.push(pii(required->first, idx + 1));
          max = (constTime[required->first] > max) ? constTime[required->first] : max;
          D(cout << "Max: " << max << endl;)
        } 
        steps.erase(required);
        required = lower_bound(steps.begin(), steps.end(), building, [](const pii& lhs, const ui rhs) {
          return lhs.second < rhs;
        });
      }
      return max;
    };

    vector<ui> maxVector(K + 1);

    while(buildingToBuild.empty() == false) {
      pii b = buildingToBuild.front();
      D(cout << "Find max of: " << b.first << endl;)
      ui max = findToBuild(b.first, b.second);
      D(cout << "Current max: " << max << endl;)
      maxVector[b.second] = max > maxVector[b.second] ? max : maxVector[b.second]; 
      buildingToBuild.pop();
    };

    for(auto m: maxVector) {
      D(cout << m << " ";)
      totalTime += m;
    }
    D(cout << endl;)

    D(cout << "total: " << totalTime << endl;)

    
    cout << totalTime << endl;
  }


  return 0;
}